src/input/dotgraph.rs

```diff
@@ -33,14 +33,70 @@ impl<Node: Idx, Label: Idx> InputCapabilities<Node> for DotGraphInput<Node, Labe
     type GraphInput = DotGraph<Node, Label>;
 }
 
-pub struct DotGraph<Node: Idx, Label: Idx> {
-    node_count: Node,
-    edge_count: Node,
-    labels: Vec<Label>,
-    edges: EdgeList<Node>,
-    max_degree: Node,
-    max_label: Label,
-    label_frequency: HashMap<Label, usize>,
+pub struct DotGraph<Node, Label>
+where
+    Node: Idx,
+    Label: Idx,
+{
+    pub(crate) labels: Vec<Label>,
+    pub(crate) edge_list: EdgeList<Node>,
+    pub(crate) max_degree: Node,
+    pub(crate) max_label: Label,
+    pub(crate) label_frequency: HashMap<Label, usize>,
+}
+
+impl<Node, Label> DotGraph<Node, Label>
+where
+    Node: Idx,
+    Label: Idx + Hash,
+{
+    fn node_count(&self) -> Node {
+        Node::new(self.labels.len())
+    }
+
+    pub(crate) fn label_count(&self) -> usize {
+        if self.label_frequency.len() > self.max_label.index() + 1 {
+            self.label_frequency.len()
+        } else {
+            self.max_label.index() + 1
+        }
+    }
+
+    pub(crate) fn max_label_frequency(&self) -> usize {
+        self.label_frequency
+            .values()
+            .max()
+            .cloned()
+            .unwrap_or_default()
+    }
+
+    pub(crate) fn label_index(&self) -> Csr<Label, Node> {
+        let node_count = self.node_count();
+        let label_count = self.label_count();
+
+        let mut nodes = vec![Node::zero(); node_count.index()];
+        let mut offsets = Vec::with_capacity(label_count.index() + 1);
+        offsets.push(Label::zero());
+
+        let mut total = Label::zero();
+
+        for label in Label::zero()..Label::new(label_count) {
+            offsets.push(total);
+            let freq = self.label_frequency.get(&label).unwrap_or(&0);
+            total += Label::new(*freq);
+        }
+
+        for (node, &label) in self.labels.iter().enumerate().take(node_count.index()) {
+            let offset = offsets[(label + Label::new(1)).index()];
+            nodes[offset.index()] = Node::new(node);
+            offsets[(label + Label::new(1)).index()] += Label::new(1);
```

Comment:
maybe just make var `offset` mut and add to that here instead?

Reply:
I think this is outdated. I replaced this with a multi-threaded impl.